content/src/controller.js

```diff
@@ -786,6 +793,87 @@ function saveAction(forceOverwrite, loginPrompt, doneCallback) {
   });
 }
 
+function generateThumbnailDataUrl() {
```

Comment:
This function will likely get more complicated as you add handling for more cases.  Consider pulling it out to its own file.